### PR TITLE
Feature navp 445 contributor frame in ptref and json

### PR DESCRIPTION
**Pull request by kadhikari**
1. Api added for contributor and frame 
2. graph on ptreferential modified for contributor and frame: vehicle_journey is linked to a frame and frame is linked to a contributor.
3. For details refer to : http://jira.canaltp.fr/browse/NAVP-445

### CODE REVIEW

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 1]
```diff
@@ -243,6 +243,12 @@ def output(self, key, obj):
             r.append({"id": note_.uri, "value": note_.note})
         return r
 
+class realtime_level(fields.Raw):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: I don't think you need this, there is already some generic code to output some enum

File: source/ed/build_helper.cpp [hunk 2]
```diff
@@ -572,6 +572,47 @@ void builder::connection(const std::string & name1, const std::string & name2, f
...
+    this->data->pt_data->contributors.push_back(contributor);
+
+    frame = new navitia::type::Frame();
+    frame->idx = this->data->pt_data->frames.size();
+    frame->uri = "f3";
+    frame->name = "name-f3";
+    frame->contributor = contributor;
+    contributor->frame_list.push_back(frame);
+    this->data->pt_data->frames.push_back(frame);
+    this->frames[frame->uri]=frame;
```
Comment (antoine-de) [inline source/ed/build_helper.cpp]: why don't you use the pt_data map ?

File: source/ed/build_helper.h [hunk 3]
```diff
@@ -180,6 +180,8 @@ struct builder {
     std::map<std::string, navitia::type::StopPoint *> sps;
     std::map<std::string, navitia::type::Network *> nts;
     std::multimap<std::string, navitia::type::VehicleJourney*> block_vjs;
+    std::map<std::string, navitia::type::Frame *> frames;
```
Comment (antoine-de) [inline source/ed/build_helper.h]: same here, I think you should use the pt_data map

File: source/ed/ed_reader.cpp [hunk 4]
```diff
@@ -886,6 +887,7 @@ void EdReader::fill_vehicle_journeys(nt::Data& data, pqxx::work& work){
             auto frame_it = this->frame_map.find(const_it["frame_id"].as<idx_t>());
             if(frame_it != this->frame_map.end()) {
                 vj->frame = frame_it->second;
+                vj->frame->vehiclejourney_list.push_back(vj);
```
Comment (antoine-de) [inline source/ed/ed_reader.cpp:890]: I would be more in favor not to stock this back reference, but to compute it at the request.

what do you think about that @TeXitoi @pbougue @xlqian @kinnou02 ?

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 5]
```diff
@@ -601,6 +607,23 @@ def output(self, key, region):
...
+contributor = {
+    "id": fields.String(attribute='uri'),
+    "name": fields.String(),
+    "website": fields.String(),
+    "license": fields.String()
+}
+
+frame = {
+    "id": fields.String(attribute='uri'),
+    "desc": fields.String(),
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: whouldn't `description` be more explicit ?

File: source/ptreferential/tests/ptref_test.cpp [hunk 6]
```diff
@@ -483,9 +483,11 @@ BOOST_AUTO_TEST_CASE(find_path_test){
     Jointures j;
     std::vector<Jointures::vertex_t> component(boost::num_vertices(j.g));
     int num = boost::connected_components(j.g, &component[0]);
-    BOOST_CHECK_EQUAL(num, 5);
+    //BOOST_CHECK_EQUAL(num, 5);
```
Comment (antoine-de) [inline source/ptreferential/tests/ptref_test.cpp]: you can remove this line (and the one below)

File: source/ed/ed_reader.cpp (hunk 4 above)
Comment (TeXitoi) [inline source/ed/ed_reader.cpp:890]: I don't care.

File: source/ptreferential/tests/ptref_test.cpp [hunk 7]
```diff
@@ -698,3 +700,84 @@ BOOST_AUTO_TEST_CASE(code_request) {
     BOOST_REQUIRE_EQUAL(res.size(), 1);
     BOOST_CHECK_EQUAL(b.data->pt_data->stop_areas.at(res.at(0))->uri, "A");
 }
+
+BOOST_AUTO_TEST_CASE(contributor_and_frame) {
+
+    ed::builder b("201601011T1739");
```
Comment (antoine-de) [inline source/ptreferential/tests/ptref_test.cpp:704]: If the default frames and contributor are only used in one test, I think it would be clearer to add them only there no ?

File: source/type/pb_converter.cpp [hunk 8]
```diff
@@ -334,15 +335,45 @@ void fill_pb_object(const georef::Admin* adm, const nt::Data&,
...
-                    const pt::time_period& , const bool, const DumpMessage){}
+void fill_pb_object(const nt::Frame* fr,
+                    const nt::Data& data, pbnavitia::Frame* frame,
+                    int depth, const pt::ptime&  now,
+                    const pt::time_period& action_period,
+                    const bool show_codes, const DumpMessage dump_message){
+    if(fr == nullptr)
+        return;
+    frame->set_uri(fr->uri);
+    frame->set_start_validation_date(navitia::to_posix_timestamp(pt::ptime(fr->validation_period.begin(),pt::time_duration(0, 0, 0, 0))));
```
Comment (antoine-de) [inline source/type/pb_converter.cpp]: minor: you can replace `pt::time_duration(0, 0, 0, 0)` by `0_s`

File: source/ed/build_helper.h (hunk 3 above)
Comment (kadhikari) [inline source/ed/build_helper.h]: I don't need this map so i deleted.

**Comment by kadhikari** [comment]
@antoine-de 
I made some modifications

File: source/ed/build_helper.cpp [hunk 9]
```diff
@@ -570,7 +570,7 @@ void builder::connection(const std::string & name1, const std::string & name2, f
     commercial_mode->idx = this->data->pt_data->commercial_modes.size();
     commercial_mode->name = "Car";
     commercial_mode->uri = "Car";
-    this->data->pt_data->commercial_modes.push_back(commercial_mode);
+    this->data->pt_data->commercial_modes.push_back(commercial_mode);
```
Comment (antoine-de) [inline source/ed/build_helper.cpp]: you have a trailing space here.

File: source/ed/ed_reader.cpp (hunk 4 above)
Comment (pbougue) [inline source/ed/ed_reader.cpp:890]: I vote for @antoine-de (even if it hurts), as it is the same problem (and roughly the same amount of requests I guess) than navitia::type::PhysicalMode::get.

File: source/type/pb_converter.cpp [hunk 10]
```diff
@@ -334,15 +335,46 @@ void fill_pb_object(const georef::Admin* adm, const nt::Data&,
...
+    }
+    if(fr->realtime_level == nt::RTLevel::RealTime){
+        frame->set_realtime_level(pbnavitia::REAL_TIME);
+    }
+
+    if(fr->contributor != nullptr) {
+        fill_pb_object(fr->contributor, data, frame->mutable_contributor(), depth-1,
+                      now, action_period, show_codes, dump_message);
+    }
+    }
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: bad indentation

File: source/ed/ed_reader.cpp (hunk 4 above)
Comment (xlqian) [inline source/ed/ed_reader.cpp:890]: will it be expensive to compute it  at the request?

File: source/type/pb_converter.cpp (hunk 10 above)
Comment (pbougue) [inline source/type/pb_converter.cpp]: minor: indent

File: source/type/pb_converter.cpp [hunk 11]
```diff
@@ -334,15 +335,46 @@ void fill_pb_object(const georef::Admin* adm, const nt::Data&,
...
+    frame->set_uri(fr->uri);
+    pt::time_duration td = pt::time_duration(0, 0, 0, 0);
+    frame->set_start_validation_date(navitia::to_posix_timestamp(pt::ptime(fr->validation_period.begin(), td)));
+    frame->set_end_validation_date(navitia::to_posix_timestamp(pt::ptime(fr->validation_period.end(),td)));
+    frame->set_desc(fr->desc);
+    frame->set_system(fr->system);
+    if(fr->realtime_level == nt::RTLevel::Base){
+        frame->set_realtime_level(pbnavitia::BASE);
+    }
+    if(fr->realtime_level == nt::RTLevel::Adapted){
```
Comment (xlqian) [inline source/type/pb_converter.cpp]: else if

File: source/type/pb_converter.cpp [hunk 12]
```diff
@@ -334,15 +335,46 @@ void fill_pb_object(const georef::Admin* adm, const nt::Data&,
...
+        frame->set_realtime_level(pbnavitia::ADAPTED);
+    }
+    if(fr->realtime_level == nt::RTLevel::RealTime){
```
Comment (xlqian) [inline source/type/pb_converter.cpp]: else if

File: source/type/pb_converter.cpp (hunk 11 above)
Comment (xlqian) [inline source/type/pb_converter.cpp]: else if

File: source/type/pb_converter.cpp (hunk 12 above)
Comment (xlqian) [inline source/type/pb_converter.cpp]: else if

File: source/ptreferential/tests/ptref_test.cpp [hunk 13]
```diff
@@ -483,9 +483,9 @@ BOOST_AUTO_TEST_CASE(find_path_test){
     Jointures j;
     std::vector<Jointures::vertex_t> component(boost::num_vertices(j.g));
     int num = boost::connected_components(j.g, &component[0]);
-    BOOST_CHECK_EQUAL(num, 5);
+    BOOST_CHECK_EQUAL(num, 3);
     num =  boost::strong_components(j.g, &component[0]);
-    BOOST_CHECK_EQUAL(num, 6);
+    BOOST_CHECK_EQUAL(num, 4);
 
     // Type qui n'existe pas dans le graph : il n'y a pas de chemin
```
Comment (pbougue) [inline source/ptreferential/tests/ptref_test.cpp:490]: minor: If you're up to translate...

File: source/type/pb_converter.cpp (hunk 11 above)
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: Better: switch

File: source/ed/ed_reader.cpp (hunk 4 above)
Comment (pbougue) [inline source/ed/ed_reader.cpp:890]: Well, it's doable (as it is already done for physical mode)

Comment (TeXitoi) [inline source/ed/ed_reader.cpp:890]: It will be much more expensive, but I don't think it will be visible. /vehicle_journeys/id does this loop.

File: source/ptreferential/tests/ptref_test.cpp (hunk 7 above)
Comment (pbougue) [inline source/ptreferential/tests/ptref_test.cpp:704]: Done

File: source/type/pb_converter.cpp (hunk 12 above)
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: better: switch

**Comment by TeXitoi** [comment]
OK for me.

File: source/ptreferential/tests/ptref_test.cpp (hunk 7 above)
Comment (TeXitoi) [inline source/ptreferential/tests/ptref_test.cpp]: It will be changed in another PR

File: source/ed/ed_reader.cpp (hunk 4 above)
Comment (TeXitoi) [inline source/ed/ed_reader.cpp:890]: The change will be done in another PR.

**Comment by pbougue** [comment]
Ok for me
